cf/read_write/netcdf/netcdfwrite.py

```diff
@@ -455,25 +643,399 @@ def _filled_string_array(self, array, fill_value=""):
+            dx_ind = tuple(range(dx.ndim))
+            out_ind = dx_ind
+            dx = da.blockwise(
+                self._cfa_unique,
+                out_ind,
+                dx,
+                dx_ind,
+                adjust_chunks={i: 1 for i in out_ind},
+                dtype=dx.dtype,
+            )
+
+            # Get the non-standard term name from the field
+            # ancillary's 'id' attribute
+            term = getattr(field_anc, "id", "term")
+            term = term.replace(" ", "_")
+            name = term
+            n = 0
+            while term in terms:
+                n += 1
+                term = f"{name}_{n}"
+
+            terms.append(term)
+
+            # Create the new CFA term variable
+            term_ncvar = self._cfa_write_term_variable(
+                data=type(data)(dx),
+                ncvar=aggregated_data.get(term, f"cfa_{term}"),
+                ncdimensions=fragment_ncdimensions,
+            )
+
+            aggregated_data_attr.append(f"{term}: {term_ncvar}")
+
+        return aggregated_data_attr
+
+    @classmethod
+    def _cfa_unique(cls, a):
+        """Return the unique value of an array.
+
+        If there are multipl unique vales then missing data is
+        returned.
+
+        .. versionadded:: TODOCFAVER
+
+        :Parameters:
+
+             a: `numpy.ndarray`
+                The array.
+
+        :Returns:
+
+            `numpy.ndarray`
+                A size 1 array containing the unique value, or missing
+                data if there is not a unique unique value.
+
+        """
+        out_shape = (1,) * a.ndim
+        a = np.unique(a)
+        if np.ma.isMA(a):
+            # Remove a masked element
+            a = a.compressed()
+
+        if a.size == 1:
+            return a.reshape(out_shape)
+
+        return np.ma.masked_all(out_shape, dtype=a.dtype)
 
-#    def _convert_dtype(self, array, new_dtype=None):
-#        """Convert the data type of a numpy array.
-#
-#        .. versionadded:: 3.14.0
-#
-#        :Parameters:
-#
-#            array: `numpy.ndarray`
-#                The `numpy` array
-#
-#            new_dtype: data-type
-#                The new data type.
-#
-#        :Returns:
-#
-#            `numpy.ndarray`
-#                The array with converted data type.
-#
-#        """
-#        return array.astype(new_dtype)
-#
+    def _cfa_aggregation_instructions(self, data, cfvar):
+        """Convert data to standardised CFA aggregation instruction terms.
+
+        .. versionadded:: TODOCFAVER
+
+        :Parameters:
+
+            data: `Data`
+                The data to be converted to standardised CFA
+                aggregation instruction terms.
+
+            cfvar: construct
+                The construct that contains the *data*.
+
+        :Returns:
+
+            `dict`
+                A dictionary whose keys are the standardised CFA
+                aggregation instruction terms, keyed by `Data`
```

Comment:
'Keyed' to me implies the keys themselves, not the values, therefore:

```suggestion
                aggregation instruction terms, with values of `Data`
```

or similar?